spatialos-sdk/src/worker/entity.rs

```diff
@@ -0,0 +1,50 @@
+use crate::worker::component::{self, Component, ComponentId};
+use spatialos_sdk_sys::worker::Worker_ComponentData;
+use std::collections::HashSet;
+use std::ptr;
+
+#[derive(Debug, Default)]
+pub struct Entity {
+    component_ids: HashSet<ComponentId>,
+    pub(crate) component_data: Vec<Worker_ComponentData>,
+    drops: Vec<unsafe fn(*mut std::ffi::c_void)>,
+}
+
+impl Entity {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn add<C: Component>(&mut self, component: C) {
+        assert!(
+            !self.component_ids.contains(&C::component_id()),
+            "Duplicate component added to `Entity`"
+        );
+
+        let data_ptr = component::handle_allocate(component);
+        let component_data = Worker_ComponentData {
+            reserved: ptr::null_mut(),
+            component_id: C::component_id(),
+            schema_type: ptr::null_mut(),
+            user_handle: data_ptr as *mut _,
+        };
+
+        self.component_ids.insert(C::component_id());
+        self.component_data.push(component_data);
+        self.drops.push(drop_raw::<C>);
+    }
```

Comment:
Could you also add the functions `pub fn get<C: Component>(&self) -> &C` and `pub fn get_mut<C: Component>(&mut self) -> &mut C`?

Reply:
I'm down to add `get`, but I'm not sure `get_mut` is safe to add. Once you pass the entity to the networking thread for serialization, it wouldn't be safe to get a mutable reference to the component data. I _think_ we can make it safe if we can ensure that the user always gives up ownership of the `Entity` when they send for serialization, but I'm not quite sure how to do that at the moment. That said, I should be able to safely add a `remove` method, which will allow the user to remove the component, modify it, and then add it back if they want.

Reply:
Hmmmm, actually removing a component also presents a challenge if the data has already been sent to the networking thread. We can remove the component from the `HashMap` and drop the `Arc` handle in order to decrease the refcount, but we won't be able to return the component data to the user if the handle has already been copied. I can have it return a `Result<C, RemoveErr>` where `RemoveErr` is an enum with variants to distinguish between "no component was in the entity" and "component removed from entity but it's being used on another thread".

Do you have any specific use case in mind for this API? I'm not sure how we're expecting it to be used, so I can't tell what design for the API makes the most sense.

Reply:
I personally think of the `Entity` class as some kind of view into a specific Entity from the perspective of the worker. In theory, we may want to maintain this view by merging component updates into it, and that would require you to be able to mutate data currently stored within the `HashMap`.

I agree that it's definitely unsafe to modify components once they've been sent to the networking thread, so I would probably suggest instead that you make a copy of all the component data when converting the `HashMap` into `Vec<Worker_ComponentData>`, rather than relying on reference counting. This also avoids the component removal issue.

Reply:
Another way to avoid the potential pitfalls of the networking thread (and avoid copying) would instead be to try and make sure there is no way for an Entity to be shared by both the application and the network threads at the same time. You can accomplish this pretty easily by only allowing the user to move an Entity object into `Connection::send_create_entity_request` (effectively consuming it).